strapi/strapi
# feat(content-manager): add possibility to customize blocks

### What does it do?

Move the `blockStore` object into the `content-manager` plugin configuration, giving the user the option to add, remove, or change `rich-text` fields.
For now I just edit the blocks part, but I believe it is necessary to adjust the `modifiers` too, but I thought it would be too much for just one PR.

### Why is it needed?

Strapi's default blocks don't suit all user cases, I made this change because I needed to add colors and I couldn't figure out how.

### How to test it?

- I tested using getstarted example, however, I did not test each case of changes to the blocks.
- The test script fails, but I saw that it is in other codes that are not even related, and when returning to the main branch it continues to fail.

====== added by reviewer ======

  1. Add `addRichTextBlocks` to any plugin's `register` hook in a Strapi app that has a content type with a Blocks field:

  ```js
  register(app) {
    const cms = app.getPlugin('content-manager').apis;

    // Add a custom block
    cms.addRichTextBlocks({
      'my-callout': {
        isInBlocksSelector: true,
        label: { id: 'my-plugin.callout', defaultMessage: 'Callout' },
        renderElement: ({ children, attributes }) => (
          <div {...attributes} style={{ borderLeft: '4px solid orange', padding: '8px 12px' }}>
            {children}
          </div>
        ),
        matchNode: (node) => node.type === 'my-callout',
        handleConvert(editor) {
          Transforms.setNodes(editor, { type: 'my-callout' });
        },
      },
    });

    // Remove a default block
    cms.addRichTextBlocks((prev) => {
      const { code, ...rest } = prev;
      return rest;
    });
  }
```
  2. Open the Blocks editor for any entry, "Callout" should appear in the block-type dropdown and "Code" should be gone.
  3. Select "Callout", the current block converts and renders with the custom style.
  4. Verify that removing "Code" doesn't crash the editor.
